---
title: "Consensus OPLS analyses"
author: "Florence Mehl"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  html_document: 
    toc: TRUE
    theme: paper
    higlight: tango
    code_folding: "hide"

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(readxl)
library(ggplot2)
library(ggrepel)
library(reshape2)
library(ConsensusOPLS)

```

# Dataset from Wigger et al. Nature Metabolism 2021

133 adults undergoing pancreatic surgery at the University Hospital Carl Gustav Carus Dresden.  
Based on the thresholds set by the American Diabetes Association4 (ADA) for fasting glucose, HbA1c and 2-hour glycaemia of an OGTT in the days immediately before surgery, 18 of these individuals were classified as ND, 41 with IGT, including 3 with IFG only, 35 with T3cD and 39 with T2D.  
A diagnosis of T3cD was made whenever the occurrence of diabetes was not recognized for longer than 1 year prior to the onset of the symptoms leading to surgery and the patient was negative for the presence of circulating autoantibodies against pancreatic islets.  
IFG and IGT patients were merged in one group that is hereafter labeled as IGT.  
Among the 133 patients, 51 patients had transcriptomics, shotgun lipidomics and targeted sphingolipids measurements. 4 patients were classified as ND, 21 patients as IGT, 12 patients as T3D and 14 patients as T2D. 

```{r import and format dataset, message=FALSE, warning=FALSE}

# Metadata
metadata <- read_xlsx("metadata_PPP.xlsx")
colnames(metadata)[2] <- "Diabetes status"
metadata[,3:6] <- apply(metadata[,3:6], 2, as.numeric)
metadata <- metadata[!is.na(metadata$HbA1c),]
rownames(metadata) <- metadata$`Patient ID`
metadata.patients <- rownames(metadata)

# RNAseq
rna.orig <- read.delim("RNASeq_vst_batch_corrected_data_LCM.txt") 
rna <- rna.orig[,-1,]
rownames(rna) <- rna.orig$X
rna <- t(rna)
rna.patients <- rownames(rna)

# Shotgun lipidomics
lip.orig <- read.delim("shotgun_lipidomics_filtered_imputed_plasma_PPP.txt")
lip.metadata <- lip.orig[,c(118:122)]
rownames(lip.metadata) <- lip.orig$pid
lip <- lip.orig[,c(-1,-118:-122)]
rownames(lip) <- lip.orig$pid
lip.patients <- rownames(lip.metadata[lip.metadata$included_in_cOPLS_analysis == 1,])

# Targeted Sphingolipids
sphingo.orig <- read.delim("targeted_sphingolipids_filtered_imputed_plasma_PPP.txt")
sphingo.metadata <- sphingo.orig[,c(16:20)]
rownames(sphingo.metadata) <- sphingo.orig$pid
sphingo <- sphingo.orig[,c(-1,-16:-20)]
rownames(sphingo) <- sphingo.orig$pid
sphingo.patients <- rownames(sphingo.metadata[sphingo.metadata$included_in_cOPLS_analysis == 1,])

# Common patients data
patients <- Reduce(intersect, list(rna.patients, lip.patients, sphingo.patients, metadata.patients))
rna <- rna[rownames(rna) %in% patients,]
lip <- lip[rownames(lip) %in% patients,]
sphingo <- sphingo[rownames(sphingo) %in% patients,]
metadata <- metadata[patients,]
rownames(metadata) <- patients
metadata <- cbind(metadata[patients, ], lip.metadata[patients, ])

diabetes.status.colors <- c("#3F80CD", "#77933C", "#810000", "#E46C0A")
names(diabetes.status.colors) <- c("ND", "IGT", "T3D", "T2D")	
```

# Consensus OPLS regression on HbA1c

```{r consensusOPLS regression, message=FALSE, warning=FALSE, fig.align='center'}

COPLS_data <- list(rna=as.matrix(rna),
                   lipids=as.matrix(lip),
                   sphingolipids = as.matrix(sphingo))
COPLS_data <- lapply(COPLS_data, scale)
Y <- as.matrix(metadata$HbA1c)

COPLS_res <- ConsensusOPLS(
  data = COPLS_data,
  Y = Y,
  maxPcomp = 1,
  maxOcomp = 1,
  nfold = 51,
  cvType = "nfold",
  nperm = 1000,
  modelType = "reg",
  mc.cores = 1,
  verbose = FALSE,
  plots = T
)

```

## Permutations

```{r reg permutations, message=FALSE, warning=FALSE, fig.align='center'}

Q2perm <- data.frame(Q2perm = COPLS_res$permStats$Q2Yhat)

ggplot(data = Q2perm, aes(x = Q2perm)) +
  geom_histogram(color="grey", fill="grey") +
  geom_vline(aes(xintercept=COPLS_res$optimal$modelCV$cv$Q2Yhat[2]),color="blue", linetype="dashed", size=1) +
  theme_classic() +
  ggtitle("Q2 Permutation test")


R2Yperm <- data.frame(R2Yperm = COPLS_res$permStats$R2Yhat)

ggplot(data = R2Yperm, aes(x = R2Yperm)) +
  geom_histogram(color="grey", fill="grey") +
  geom_vline(aes(xintercept=COPLS_res$optimal$modelCV$Model$R2Yhat[2]),color="blue", linetype="dashed", size=1) +
  theme_classic() +
  ggtitle("R2Y Permutation test")

```

## Blocks contribution

```{r consensus OPLS-reg contributions, message=FALSE, warning=FALSE, fig.align='center'}

contributions <- COPLS_res$optimal$modelCV$Model$blockContribution
contributions <- melt(contributions)
colnames(contributions) <- c("dataset", "Dim", "value")

ggplot(contributions, aes(x=Dim, y=value, fill=dataset)) +
  geom_bar(stat = "identity", position=position_dodge()) +
  theme_light() +
  labs(x = "global components", y = "specific weights of block on global components", fill = "omic",
       title = "contributions")

```

## Scores Plot

```{r consensus OPLS-reg scores, message=FALSE, warning=FALSE, fig.align='center'}

scores <- data.frame(metadata, COPLS_res$optimal$modelCV$Model$scores)

ggplot(scores, aes(x=p_1, y=o_1, fill=HbA1c, col = Diabetes.status)) +
  geom_point(size=4, shape = 21, stroke = 1.5) +
  labs(x="Predictive",
       y="Orthogonal",
       title = "scores plots on predictive vs orthogonal latent variables") +
  scale_fill_gradient(low = "white", high = "black") +
  scale_color_manual(values= diabetes.status.colors) +
  theme_light()

YpredYobs <- data.frame(Yobs = Y, Ypred =COPLS_res$optimal$modelCV$cv$AllYhat[,2], metadata)

ggplot(YpredYobs) +
  geom_point(size=4, shape = 21, stroke = 1.5, aes(x=Yobs, y=Ypred, fill=HbA1c, col = Diabetes.status)) +
  labs(x="observed HbA1c",
       y="predicted HbA1c",
       title = "predicted Y vs observed Y") +
  scale_fill_gradient(low = "white", high = "black") +
  scale_color_manual(values= diabetes.status.colors) +
  geom_smooth(aes(x=Yobs, y=Ypred),method=lm, se=FALSE, col="grey") +
  theme_light()
```

## Loadings 

```{r consensus OPLS-reg loadings, message=FALSE, warning=FALSE, fig.align='center'}

loadings <- rbind.data.frame(COPLS_res$optimal$modelCV$Model$loadings$rna, COPLS_res$optimal$modelCV$Model$loadings$lipids, COPLS_res$optimal$modelCV$Model$loadings$sphingolipids)
loadings$dataset <- c(rep("genes", nrow(COPLS_res$optimal$modelCV$Model$loadings$rna)),
                      rep("lipids", nrow(COPLS_res$optimal$modelCV$Model$loadings$lipids)),
                      rep("sphingolipids", nrow(COPLS_res$optimal$modelCV$Model$loadings$sphingolipids)))
loadings$variable <- rownames(loadings)

ggplot(loadings, aes(x=p_1, y=o_1, col=dataset, label = variable)) +
  geom_point(size=2, alpha = 0.5) +
  labs(x="Predictive",
       y="Orthogonal",
       title = "loadings plots on predictive vs orthogonal latent variables") +
  geom_text_repel(size=2, max.overlaps = 200, segment.size=.1) +
  theme_light()

```

## loadings and VIP of the optimal model

```{r consensus OPLS-reg VIP loadings, message=FALSE, warning=FALSE, fig.align='center'}

VIP <- data.frame(VIP = c(COPLS_res$optimal$VIP$rna, COPLS_res$optimal$VIP$lipids, COPLS_res$optimal$VIP$sphingolipids), variable = c(names(COPLS_res$optimal$VIP$rna), names(COPLS_res$optimal$VIP$lipids), names(COPLS_res$optimal$VIP$sphingolipids)))

loadings_VIP <- merge(loadings, VIP, by="variable")
loadings_VIP$label <- ifelse(loadings_VIP$VIP > 1, loadings_VIP$variable, NA)

ggplot(loadings_VIP, aes(x=p_1, y=VIP, col=dataset, label = label)) +
  geom_point(size=2, alpha = 0.5) +
  labs(x="Predictive",
       y="VIP",
       title = "loadings plots on predictive vs orthogonal latent variables") +
  geom_text_repel(size=3, max.overlaps = 50, segment.size=.1) +
  theme_light()

loadings_VIP <- loadings_VIP[order(loadings_VIP$VIP, decreasing = T),]
loadings_VIP$rank <- seq(1, nrow(loadings_VIP))
loadings_VIP <- loadings_VIP[loadings_VIP$VIP > quantile(loadings_VIP$VIP, 0.75, names = F),]

ggplot(loadings_VIP, aes(x=rank, y=VIP, fill=dataset, label = label)) +
  geom_bar(stat = "identity") +
  labs(x="Variable",
       y="VIP",
       title = "VIP plot") +
  theme_light()

```


# Consensus OPLS discriminant analysis of the 3 diabetes status (IGT, T2D and T3D)

```{r consensusOPLS DA, message=FALSE, warning=FALSE, fig.align='center'}

diab.patients <- metadata$`Patient ID`[metadata$`Diabetes status` %in% c("IGT", "T2D", "T3D")]
metadata.diab <- metadata[diab.patients,]

COPLS_data_DA <- list(rna=as.matrix(rna[diab.patients,]),
                      lipids=as.matrix(lip[diab.patients,]),
                      sphingolipids = as.matrix(sphingo[diab.patients,]))
COPLS_data_DA$lipids <- apply(COPLS_data_DA$lipids, 2, as.numeric)
COPLS_data_DA <- lapply(COPLS_data_DA, scale)

Y_DA <- as.matrix(data.frame(IGT = ifelse(metadata.diab$`Diabetes status` == "IGT", 1, 0),
                             T2D = ifelse(metadata.diab$`Diabetes status` == "T2D", 1, 0),
                             T3D = ifelse(metadata.diab$`Diabetes status` == "T3D", 1, 0)))

rownames(Y_DA) <- diab.patients

COPLS_res_DA <- ConsensusOPLS(
  data = COPLS_data_DA,
  Y = Y_DA,
  maxPcomp = 2,
  maxOcomp = 3,
  nfold = 47,
  cvType = "nfold",
  nperm = 1000,
  modelType = "da",
  mc.cores = 1,
  verbose = FALSE,
  plots = T
)

```

## Permutations

```{r DA permutations, message=FALSE, warning=FALSE, fig.align='center'}

Q2perm_DA <- data.frame(Q2perm = COPLS_res_DA$permStats$Q2Yhat)

ggplot(data = Q2perm_DA, aes(x = Q2perm)) +
  geom_histogram(color="grey", fill="grey") +
  geom_vline(aes(xintercept=COPLS_res_DA$optimal$modelCV$cv$Q2Yhat[2]),color="blue", linetype="dashed", size=1) +
  theme_classic() +
  ggtitle("Q2 Permutation test")

dQ2perm_DA <- data.frame(dQ2perm = COPLS_res_DA$permStats$DQ2Yhat)

ggplot(data = dQ2perm_DA, aes(x = dQ2perm)) +
  geom_histogram(color="grey", fill="grey") +
  geom_vline(aes(xintercept=COPLS_res_DA$optimal$modelCV$cv$DQ2Yhat[2]),color="blue", linetype="dashed", size=1) +
  theme_classic() +
  ggtitle("dQ2 Permutation test")


R2Yperm_DA <- data.frame(R2Yperm_DA = COPLS_res_DA$permStats$R2Yhat)

ggplot(data = R2Yperm_DA, aes(x = R2Yperm_DA)) +
  geom_histogram(color="grey", fill="grey") +
  geom_vline(aes(xintercept=COPLS_res_DA$optimal$modelCV$Model$R2Yhat[2]),color="blue", linetype="dashed", size=1) +
  theme_classic() +
  ggtitle("R2Y Permutation test")

```

## Blocks contribution


```{r consensus OPLS-DA contributions, message=FALSE, warning=FALSE, fig.align='center'}

contributions_DA <- COPLS_res_DA$optimal$modelCV$Model$blockContribution
contributions_DA <- melt(contributions_DA)
colnames(contributions_DA) <- c("dataset", "Dim", "value")

ggplot(contributions_DA, aes(x=Dim, y=value, fill=dataset)) +
  geom_bar(stat = "identity", position=position_dodge()) +
  theme_light() +
  labs(x = "global components", y = "specific weights of block on global components", fill = "omic",
       title = "contributions")

```

## Scores Plot

```{r consensus OPLS-DA scores, message=FALSE, warning=FALSE, fig.align='center'}

scores_DA <- data.frame(metadata.diab, COPLS_res_DA$optimal$modelCV$Model$scores)

ggplot(scores_DA, aes(x=p_1, y=p_2, col = Diabetes.status)) +
  geom_point(size=4) +
  labs(x="Predictive 1",
       y="Predictive 2",
       title = "scores plots on predictive latent variables") +
  scale_color_manual(values= diabetes.status.colors) +
  theme_light()

```

## Loadings 

```{r consensus OPLS-DA loadings, message=FALSE, warning=FALSE, fig.align='center'}

loadings_DA <- rbind.data.frame(COPLS_res_DA$optimal$modelCV$Model$loadings$rna, COPLS_res_DA$optimal$modelCV$Model$loadings$lipids, COPLS_res_DA$optimal$modelCV$Model$loadings$sphingolipids)
loadings_DA$dataset <- c(rep("genes", nrow(COPLS_res_DA$optimal$modelCV$Model$loadings$rna)),
                      rep("lipids", nrow(COPLS_res_DA$optimal$modelCV$Model$loadings$lipids)),
                      rep("sphingolipids", nrow(COPLS_res_DA$optimal$modelCV$Model$loadings$sphingolipids)))
loadings_DA$variable <- rownames(loadings_DA)

ggplot(loadings_DA, aes(x=p_1, y=o_1, col=dataset, label = variable)) +
  geom_point(size=2, alpha = 0.5) +
  labs(x="Predictive 1",
       y="Predictive 2",
       title = "loadings plots on predictive latent variables") +
  geom_text_repel(size=2, max.overlaps = 100, segment.size=.1) +
  theme_light()

```

## VIP of the optimal model

```{r consensus OPLS-DA VIP loadings, message=FALSE, warning=FALSE, fig.align='center'}

VIP_DA <- data.frame(VIP = c(COPLS_res_DA$optimal$VIP$rna, COPLS_res_DA$optimal$VIP$lipids, COPLS_res_DA$optimal$VIP$sphingolipids), variable = c(names(COPLS_res_DA$optimal$VIP$rna), names(COPLS_res_DA$optimal$VIP$lipids), names(COPLS_res_DA$optimal$VIP$sphingolipids)))

loadings_VIP_DA <- merge(loadings_DA, VIP_DA, by="variable")
loadings_VIP_DA$label <- ifelse(loadings_VIP_DA$VIP > 1, loadings_VIP_DA$variable, NA)
loadings_VIP_DA <- loadings_VIP_DA[order(loadings_VIP_DA$VIP, decreasing = T),]
loadings_VIP_DA$rank <- seq(1, nrow(loadings_VIP_DA))
loadings_VIP_DA <- loadings_VIP_DA[loadings_VIP_DA$VIP > quantile(loadings_VIP_DA$VIP, 0.75, names = F),]

ggplot(loadings_VIP_DA, aes(x=rank, y=VIP, fill=dataset, label = label)) +
  geom_bar(stat = "identity") +
  labs(x="Variable",
       y="VIP",
       title = "VIP plot") +
  theme_light()

```




