---
title: "Application of ConsensusOPLS package on Wigger2021 data."
author: "Celine Bougel"
date: "`r format(Sys.time(), '%B %d, %Y')`"
bibliography: 20230901934_MTH2.0WP1.4_step2_validation.bib
biblio-style: apalike
link-citations: yes
nocite: '@*'
output:
  html_document:
    code_folding: show
    df_print: kable
    highlight: pygments
    number_sections: yes
    self_contained: yes
    theme: journal
    toc: yes
    toc_depth: 4
    toc_float:
      collapsed: true
      smooth_scroll: true
  editor_options:
    chunk_output_type: console
---

<style>
body {
text-align: justify}
</style>

\tableofcontents
\listoffigures
\listoftables
\newpage

# R environment preparation

```{r setup, class.source='fold-hide'}
#install.packages("knitr")
library(knitr)
knitr::opts_chunk$set(echo = TRUE)

# To ensure repeatability
set.seed(12)
```

```{r access}
path_data_brut <- "../../../data/data_brut/Wigger2021" 
path_data_produced <- "../../../data/data_produced/R_software" 
path_results <- "../../../results" 
```

```{r packages_installation, class.source='fold-hide', warning=FALSE, include=FALSE, message=FALSE}
# install.packages(c("readxl", "ggplot2", "ggrepel", "reshape2", "ConsensusOPLS"))
# update.packages(ask = FALSE)
```

```{r packages_load, warning=FALSE, message=FALSE, class.source='fold-hide'}
library(readxl)
library(ggplot2)
library(ggrepel)
library(reshape2)
library(ConsensusOPLS)
```

```{r theme_ggplot2, class.source='fold-hide'}
theme_graphs <- theme_bw() + theme(strip.text = element_text(size=14),
                                   axis.title = element_text(size=16),
                                   axis.text = element_text(size=14),
                                   plot.title = element_text(size = 16),
                                   legend.title = element_text(size = 14))
```

# QR code

```{r QRcode_Matlab}
# path_logo <- "../../../communication/presentations/Logos/" 
# 
# #install.packages("qrcode")
# library(qrcode)
# #install.packages("magick")
# library(magick)
# 
# # Link
# url_matlab <- "https://gitlab.unige.ch/Julien.Boccard/consensusopls"
# qr <- qrcode::qr_code(url_matlab, ecl = "L")
# plot(qr)
# 
# # Convertir le QR code en image magick
# qr_img <- image_graph(width = 10*ncol(qr), height = 10*nrow(qr))
# grid::grid.raster(qr, interpolate = FALSE)
# dev.off()
# 
# # Inverser les couleurs: faire le fond transparent et le motif coloré
# qr_inverted <- image_negate(qr_img) %>%
#                image_transparent("black") %>%
#                image_colorize(100, "chocolate2")
# 
# white_bg <- image_blank(image_info(qr_inverted)$width,
#                         image_info(qr_inverted)$height, 
#                         color = "white")
# 
# # Superposer l'image blanche sur le QR code inverti
# qr_inverted <- image_composite(white_bg, qr_inverted, operator = "over")
#     
# 
# # Lire le logo
# logo <- image_read(paste0(path_logo, "Matlablogo.jpg"))
# 
# #Ajouter fond blanc
# logo_with_bg <- logo #image_background(logo, "white", flatten = TRUE)
# 
# 
# # Redimensionner le logo pour qu'il soit adapté au centre du QR code
# qr_size <- min(10 * ncol(qr), 10 * nrow(qr))
# logo_size <- round(qr_size * 0.15)  # 15% de la taille du QR code
# logo_resized <- image_resize(logo_with_bg, 
#                              geometry = paste0(logo_size, "x", logo_size))
# 
# # Ajouter le logo au centre du QR code
# qr_with_logo <- image_composite(qr_inverted, logo_resized, 
#                                 operator = "atop", gravity = "center")
#                                 #operator = "over", gravity = "center")
# 
# # Sauvegarder l'image finale
# image_write(qr_inverted, 
#             path = paste0(path_logo, "/", "QRcode_Matlab_SansLogo.png"), 
#             format = "png")
# image_write(qr_with_logo, 
#             path = paste0(path_logo, "/", "QRcode_Matlab_logo.png"), 
#             format = "png")
```

```{r QRcode_R}
# # Link
# url_R <- "https://CRAN.R-project.org/package=ConsensusOPLS"
# qr <- qrcode::qr_code(url_R, ecl = "L")
# plot(qr)
# 
# # Convertir le QR code en image magick
# qr_img <- image_graph(width = 10*ncol(qr), height = 10*nrow(qr))
# grid::grid.raster(qr, interpolate = FALSE)
# dev.off()
# 
# # Inverser les couleurs: faire le fond transparent et le motif coloré
# qr_inverted <- image_negate(qr_img) %>%
#                image_transparent("black") %>%
#                image_colorize(100, "lightskyblue3")
# 
# white_bg <- image_blank(image_info(qr_inverted)$width,
#                         image_info(qr_inverted)$height,
#                         color = "white")
# 
# # Superposer l'image blanche sur le QR code inverti
# qr_inverted <- image_composite(white_bg, qr_inverted, operator = "over")
# 
# 
# # Lire le logo
# logo <- image_read(paste0(path_logo, "Rlogo.png"))
# 
# #Ajouter fond blanc
# logo_with_bg <- logo #image_background(logo, "white", flatten = TRUE)
# 
# 
# # Redimensionner le logo pour qu'il soit adapté au centre du QR code
# qr_size <- min(10 * ncol(qr), 10 * nrow(qr))
# logo_size <- round(qr_size * 0.15)  # 15% de la taille du QR code
# logo_resized <- image_resize(logo_with_bg,
#                              geometry = paste0(logo_size, "x", logo_size))
# 
# # Ajouter le logo au centre du QR code
# qr_with_logo <- image_composite(qr_inverted, logo_resized,
#                                 operator = "atop", gravity = "center")
#                                 #operator = "over", gravity = "center")
# 
# # Sauvegarder l'image finale
# image_write(qr_inverted,
#             path = paste0(path_logo, "/", "QRcode_R_SansLogo.png"),
#             format = "png")
# image_write(qr_with_logo,
#             path = paste0(path_logo, "/", "QRcode_R_logo.png"),
#             format = "png")
```

```{r QRcode_theo_rfmf_excel}
# # Link
# url_excel <- "https://docs.google.com/spreadsheets/d/e/2PACX-1vQbpa2OjCNfkn3i12OlltlhdJlWG8Xz6FT2RktRG6kI2-gLWUI2zReqaoe7gknoww/pubhtml"
# qr <- qrcode::qr_code(url_excel, ecl = "L")
# 
# # Convertir le QR code en image magick
# qr_img <- image_graph(width = 10*ncol(qr), height = 10*nrow(qr))
# grid::grid.raster(qr, interpolate = FALSE)
# dev.off()
# 
# # Inverser les couleurs: faire le fond transparent et le motif coloré
# qr_inverted <- image_negate(qr_img) %>%
#                image_transparent("black") %>%
#                image_colorize(100, "forestgreen")
# 
# white_bg <- image_blank(image_info(qr_inverted)$width, image_info(qr_inverted)$height, color = "white")
# 
# # Superposer l'image blanche sur le QR code inverti
# qr_inverted <- image_composite(white_bg, qr_inverted, operator = "over")
#     
# 
# # Lire le logo
# logo <- image_read(paste0(path_logo, "Logo_Excel.png"))
# 
# #Ajouter fond blanc
# logo_with_bg <- logo #image_background(logo, "white", flatten = TRUE)
# 
# 
# # Redimensionner le logo pour qu'il soit adapté au centre du QR code
# qr_size <- min(10 * ncol(qr), 10 * nrow(qr))
# logo_size <- round(qr_size * 0.15)  # 15% de la taille du QR code
# logo_resized <- image_resize(logo_with_bg, 
#                              geometry = paste0(logo_size, "x", logo_size))
# 
# # Ajouter le logo au centre du QR code
# qr_with_logo <- image_composite(qr_inverted, logo_resized, 
#                                 operator = "atop", gravity = "center")
#                                 #operator = "over", gravity = "center")
# 
# # Sauvegarder l'image finale
# image_write(qr_inverted, 
#             path = paste0(path_logo, "QRcode_Excel_SansLogo.png"), 
#             format = "png")
# image_write(qr_with_logo, 
#             path = paste0(path_logo, "QRcode_Excel.png"), 
#             format = "png")
```

```{r QRcode_theo_rfmf_peak_forest}
# # Link
# url_peak <- "https://demo.peakforest.org/webapp/"
# qr <- qrcode::qr_code(url_peak, ecl = "L")
# 
# # Convertir le QR code en image magick
# qr_img <- image_graph(width = 10*ncol(qr), height = 10*nrow(qr))
# grid::grid.raster(qr, interpolate = FALSE)
# dev.off()
# 
# # Inverser les couleurs: faire le fond transparent et le motif coloré
# qr_inverted <- image_negate(qr_img) %>%
#                image_transparent("black") %>%
#                image_colorize(100, "dodgerblue3")
# 
# white_bg <- image_blank(image_info(qr_inverted)$width, image_info(qr_inverted)$height, color = "white")
# 
# # Superposer l'image blanche sur le QR code inverti
# qr_inverted <- image_composite(white_bg, qr_inverted, operator = "over")
#     
# 
# # Lire le logo
# logo <- image_read(paste0(path_logo, "Logo_Peak_Forest.png"))
# 
# #Ajouter fond blanc
# logo_with_bg <- image_background(logo, "white", flatten = TRUE)
# 
# 
# # Redimensionner le logo pour qu'il soit adapté au centre du QR code
# qr_size <- min(10 * ncol(qr), 10 * nrow(qr))
# logo_size <- round(qr_size * 0.15)  # 15% de la taille du QR code
# logo_resized <- image_resize(logo_with_bg, 
#                              geometry = paste0(logo_size, "x", logo_size))
# 
# # Ajouter le logo au centre du QR code
# qr_with_logo <- image_composite(qr_inverted, logo_resized, 
#                                 operator = "over", gravity = "center")
#                                 #operator = "atop", gravity = "center")
# 
# # Sauvegarder l'image finale
# image_write(qr_inverted, 
#             path = paste0(path_logo, "QRcode_Peak_Forest_SansLogo.png"), 
#             format = "png")
# image_write(qr_with_logo, 
#             path = paste0(path_logo, "QRcode_Peak_Forest.png"), 
#             format = "png")
```

```{r QRcode_baptiste_rfmf_video1}
# # Ficher excel
# url_PCA <- "https://drive.google.com/file/d/1scrbN29pTb99fR3ukSkG1oiDcjVj9PSo/view?usp=drive_link"
# qr <- qrcode::qr_code(url_PCA, ecl = "L")
# 
# # Convertir le QR code en image magick
# qr_img <- image_graph(width = 10*ncol(qr), height = 10*nrow(qr))
# grid::grid.raster(qr, interpolate = FALSE)
# dev.off()
# 
# # Inverser les couleurs: faire le fond transparent et le motif coloré
# qr_inverted <- image_negate(qr_img) %>%
#                image_transparent("black") %>%
#                image_colorize(100, "darkolivegreen4")
# 
# white_bg <- image_blank(image_info(qr_inverted)$width, image_info(qr_inverted)$height, color = "white")
# 
# # Superposer l'image blanche sur le QR code inverti
# qr_inverted <- image_composite(white_bg, qr_inverted, operator = "over")
#     
# 
# # Lire le logo
# logo <- image_read(paste0(path_logo, "image_QR-removebg-preview.png"))
# 
# #Ajouter fond blanc
# logo_with_bg <- image_background(logo, "white", flatten = TRUE)
# 
# 
# # Redimensionner le logo pour qu'il soit adapté au centre du QR code
# qr_size <- min(10 * ncol(qr), 10 * nrow(qr))
# logo_size <- round(qr_size * 0.15)  # 30% de la taille du QR code
# logo_resized <- image_resize(logo_with_bg, 
#                              geometry = paste0(logo_size, "x", logo_size))
# 
# # Ajouter le logo au centre du QR code
# qr_with_logo <- image_composite(qr_inverted, logo_resized, 
#                                 operator = "over", gravity = "center")
#                                 #operator = "atop", gravity = "center")
# 
# # Sauvegarder l'image finale
# image_write(qr_inverted, 
#             path = paste0(path_logo, "QRcode_Baptiste_sansLogo.png"), 
#             format = "png")
# image_write(qr_with_logo, 
#             path = paste0(path_logo, "QRcode_Baptiste.png"), 
#             format = "png")
```

```{r QRcode_baptiste_rfmf_video2}
# # Ficher excel
# url_raman <- "https://drive.google.com/file/d/1OmodWSV-_Bfc_dtHtpd5KwWPc02LFsdo/view?usp=drive_link"
# qr <- qrcode::qr_code(url_raman, ecl = "L")
# 
# # Convertir le QR code en image magick
# qr_img <- image_graph(width = 10*ncol(qr), height = 10*nrow(qr))
# grid::grid.raster(qr, interpolate = FALSE)
# dev.off()
# 
# # Inverser les couleurs: faire le fond transparent et le motif coloré
# qr_inverted <- image_negate(qr_img) %>%
#                image_transparent("black") %>%
#                image_colorize(100, "darkolivegreen4")
# 
# white_bg <- image_blank(image_info(qr_inverted)$width, image_info(qr_inverted)$height, color = "white")
# 
# # Superposer l'image blanche sur le QR code inverti
# qr_inverted <- image_composite(white_bg, qr_inverted, operator = "over")
#     
# 
# # Lire le logo
# logo <- image_read(paste0(path_logo, "raman.png"))
# 
# #Ajouter fond blanc
# logo_with_bg <- image_background(logo, "white", flatten = TRUE)
# 
# 
# # Redimensionner le logo pour qu'il soit adapté au centre du QR code
# qr_size <- min(10 * ncol(qr), 10 * nrow(qr))
# logo_size <- round(qr_size * 0.15)  # 30% de la taille du QR code
# logo_resized <- image_resize(logo_with_bg, 
#                              geometry = paste0(logo_size, "x", logo_size))
# 
# # Ajouter le logo au centre du QR code
# qr_with_logo <- image_composite(qr_inverted, logo_resized, 
#                                 operator = "over", gravity = "center")
#                                 #operator = "atop", gravity = "center")
# 
# # Sauvegarder l'image finale
# image_write(qr_inverted, 
#             path = paste0(path_logo, "QRcode_Baptiste_2_sansLogo.png"), 
#             format = "png")
# image_write(qr_with_logo, 
#             path = paste0(path_logo, "QRcode_Baptiste_2.png"), 
#             format = "png")
```



# Dataset from Wigger et al. Nature Metabolism 2021

- 133 adults undergoing pancreatic surgery at the University Hospital Carl 
Gustav Carus Dresden.  

- Based on the thresholds set by the American Diabetes Association4 (ADA) for 
fasting glucose, HbA1c and 2-hour glycaemia of an OGTT in the days immediately 
before surgery, 18 of these individuals were classified as ND, 41 with IGT, 
including 3 with IFG only, 35 with T3cD and 39 with T2D.  

A diagnosis of T3cD was made whenever the occurrence of diabetes was not 
recognized for longer than 1 year prior to the onset of the symptoms leading to 
surgery and the patient was negative for the presence of circulating 
autoantibodies against pancreatic islets.  

IFG and IGT patients were merged in one group that is hereafter labeled as IGT.  

Among the 133 patients, 51 patients had transcriptomics, shotgun lipidomics and 
targeted sphingolipids measurements. 4 patients were classified as ND, 21 
patients as IGT, 12 patients as T3D and 14 patients as T2D. 

```{r import and format dataset, message=FALSE, warning=FALSE}
# Metadata
metadata <- read_xlsx(paste0(path_data_brut, "/", "metadata_PPP.xlsx"))
colnames(metadata)[2] <- "Diabetes status"
metadata[,3:6] <- apply(metadata[,3:6], 2, as.numeric)
metadata <- metadata[!is.na(metadata$HbA1c),]
rownames(metadata) <- metadata$`Patient ID`
metadata.patients <- rownames(metadata)

# RNAseq
rna.orig <- read.delim(paste0(path_data_brut, "/", "RNASeq_vst_batch_corrected_data_LCM.txt"))
rna <- rna.orig[,-1,]
rownames(rna) <- rna.orig$X
rna <- t(rna)
rna.patients <- rownames(rna)

# Shotgun lipidomics
lip.orig <- read.delim(paste0(path_data_brut, "/", "shotgun_lipidomics_filtered_imputed_plasma_PPP.txt"))
lip.metadata <- lip.orig[,c(118:122)]
rownames(lip.metadata) <- lip.orig$pid
lip <- lip.orig[,c(-1,-118:-122)]
rownames(lip) <- lip.orig$pid
lip.patients <- rownames(lip.metadata[lip.metadata$included_in_cOPLS_analysis == 1,])

# Targeted Sphingolipids
sphingo.orig <- read.delim(paste0(path_data_brut, "/", "targeted_sphingolipids_filtered_imputed_plasma_PPP.txt"))
sphingo.metadata <- sphingo.orig[,c(16:20)]
rownames(sphingo.metadata) <- sphingo.orig$pid
sphingo <- sphingo.orig[,c(-1,-16:-20)]
rownames(sphingo) <- sphingo.orig$pid
sphingo.patients <- rownames(sphingo.metadata[sphingo.metadata$included_in_cOPLS_analysis == 1,])

# Common patients data
patients <- Reduce(intersect, list(rna.patients, lip.patients, sphingo.patients, metadata.patients))
rna <- rna[rownames(rna) %in% patients,]
lip <- lip[rownames(lip) %in% patients,]
sphingo <- sphingo[rownames(sphingo) %in% patients,]
metadata <- metadata[patients,]
rownames(metadata) <- patients
metadata <- cbind(metadata[patients, ], lip.metadata[patients, ])

diabetes.status.colors <- c("#88CCEE", "#CC6677", "#DDCC77", "#117733")
omics.colors <- c("#E41A1C", "#377EB8", "#4DAF4A")
names(diabetes.status.colors) <- c("ND", "IGT", "T3D", "T2D")	
```

# Global graphs

```{r}
data <- metadata
ordre_precis <- c("ND", "IGT", "T3D", "T2D")
data$`Diabetes status` <- factor(data$`Diabetes status`, levels = ordre_precis)
data <- data[order(data$`Diabetes status`), ]


ggplot2::ggplot(data, 
                aes(x=data$`Diabetes status`,
                    y=data$HbA1c)) +
  ggplot2::ylab("HbA1c level (%)") + ggplot2::xlab(paste0("Diabetes status")) +
  ggplot2::geom_jitter(aes(col = data$`Diabetes status`,
                           group = data$`Diabetes status`),
                       position = position_jitterdodge(jitter.width = 0.2,
                                                       dodge.width = 0.8),
                       alpha = 0.2) +
  ggplot2::geom_boxplot(aes(fill = data$`Diabetes status`),
                        alpha = 0.5, outlier.shape = NA) +
  ggplot2::stat_summary(aes(group = data$`Diabetes status`),
                        position = position_dodge(width = 0.8),
                        fun = "mean", geom = "point", shape = 4,
                        size = 4, color = "red", fill = "red", na.rm = TRUE) +
  #ggpubr::stat_compare_means(method = "t.test") +
  ggplot2::scale_fill_manual(name = "Diabetes status",
                             #labels = c("lvl1", "lvl2", "lvl3", "lvl4"),
                             values = diabetes.status.colors) +
  ggplot2::scale_color_manual(name = "Diabetes status",
                              #labels = c("lvl1", "lvl2", "lvl3", "lvl4"),
                              values = diabetes.status.colors) +
  #ggplot2::scale_x_discrete(labels = c("lvl1", "lvl2", "lvl3", "lvl4")) +
  theme_graphs
```

```{r}
ggplot2::ggplot(data, 
                aes(x=data$`Diabetes status`,
                    fill=data$`Diabetes status`)) +
  ggplot2::ylab("Count") + ggplot2::xlab(paste0("Diabetes status")) +
  ggplot2::geom_bar(position = position_dodge()) +
  ggplot2::scale_fill_manual(name = "Diabetes status",
                             #labels = c("lvl1", "lvl2", "lvl3", "lvl4"),
                             values = diabetes.status.colors) +
  theme_graphs
```



# Consensus OPLS regression on HbA1c

```{r consensusOPLS regression, message=FALSE, warning=FALSE, fig.align='center'}
COPLS_data <- list(rna=as.matrix(rna),
                   lipids=as.matrix(lip),
                   sphingolipids = as.matrix(sphingo))
COPLS_data <- lapply(COPLS_data, scale)
Y <- as.matrix(metadata$HbA1c)

nbrBlocs <- length(COPLS_data)

COPLS_res <- ConsensusOPLS(
  data = COPLS_data,
  Y = Y,
  maxPcomp = 1,
  maxOcomp = 1,
  modelType = "reg",
  nperm = 1000,
  cvType = "nfold",
  nfold = 51,
  kernelParams = list(type = "p", 
                    params = c(order = 1)),
  mc.cores = 1,
  verbose = TRUE
)
```

## Permutations

```{r reg permutations, message=FALSE, warning=FALSE, fig.align='center'}
Q2perm <- data.frame(Q2perm = COPLS_res@permStats$Q2Y)

ggplot(data = Q2perm, aes(x = Q2perm)) +
  geom_histogram(color="grey", fill="grey") +
  geom_vline(aes(xintercept=COPLS_res@Q2[2]),color="blue", linetype="dashed", size=1) +
  theme_graphs +
  ggtitle("Q2 Permutation test")


R2Yperm <- data.frame(R2Yperm = COPLS_res@permStats$R2Y)

ggplot(data = R2Yperm, aes(x = R2Yperm)) +
  geom_histogram(color="grey", fill="grey") +
  geom_vline(aes(xintercept=COPLS_res@R2Y[2]),color="blue", linetype="dashed", size=1) +
  theme_graphs +
  ggtitle("R2Y Permutation test")

```

## Blocks contribution

```{r consensus OPLS-reg contributions, message=FALSE, warning=FALSE, fig.align='center'}
# contributions <- COPLS_res@blockContribution
# contributions <- melt(contributions)
# colnames(contributions) <- c("dataset", "Dim", "value")
# 
# ggplot(contributions, aes(x=Dim, y=value, fill=dataset)) +
#   geom_bar(stat = "identity", position=position_dodge()) +
#   labs(x = "Global components", y = "Specific weights of block \n on global components", fill = "omic",
#        title = "Blocks contributions") +
#   scale_x_discrete(labels = c("Pred #1", "Ortho #1")) +
#   ggplot2::scale_fill_manual(values= omics.colors)+
#   theme_graphs


data_two_plots <- data.frame("Values" = COPLS_res@blockContribution[, "p_1"],
                             "Type" = "Pred",
                             "Blocks" = labels(COPLS_data))
data_two_plots <- data.frame("Values" = c(data_two_plots$Values,
                                          COPLS_res@blockContribution[, "o_1"]),
                             "Type" = c(data_two_plots$Type,
                                        rep("Ortho", times = length(COPLS_res@blockContribution[, "o_1"]))),
                             "Blocks" = c(data_two_plots$Blocks,
                                          labels(COPLS_data)))
omics.colors <- c("#E41A1C", "#377EB8", "#4DAF4A")
ggplot(data = data_two_plots,
       aes(x = Blocks, 
           y = Values, 
           fill = Blocks)) +
  geom_bar(stat = 'identity') +
  geom_text(aes(label=round(Values, 2)), vjust=-0.3) +
  ggtitle("Block contributions to each component") +
  xlab("Components") +
  ylab("Weight") +
  facet_wrap(. ~ factor(Type)) +
  theme_graphs +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        plot.title = element_text(hjust = 0.5, 
                                  margin = margin(t = 5, r = 0, b = 0, l = 100))) +
  ggplot2::scale_fill_manual(values= omics.colors)
```

## Scores Plot

```{r consensus OPLS-reg scores, message=FALSE, warning=FALSE, fig.align='center'}
scores <- data.frame(metadata, COPLS_res@scores)

ggplot(scores, aes(x=p_1, y=o_1, fill=HbA1c, col = Diabetes.status)) +
  geom_point(size=4, shape = 21, stroke = 1.5) +
  labs(x="Predictive",
       y="Orthogonal",
       title = "Scores plots on predictive vs orthogonal latent variables") +
  scale_fill_gradient(low = "white", high = "black") +
  scale_color_manual(values= diabetes.status.colors) +
  theme_graphs

YpredYobs <- data.frame(Yobs = Y, Ypred =COPLS_res@cv$AllYhat[,2], metadata)

ggplot(YpredYobs) +
  geom_point(size=4, shape = 21, stroke = 1.5, aes(x=Yobs, y=Ypred, fill=HbA1c, col = Diabetes.status)) +
  labs(x="observed HbA1c",
       y="predicted HbA1c",
       title = "predicted Y vs observed Y") +
  scale_fill_gradient(low = "white", high = "black") +
  scale_color_manual(values= diabetes.status.colors) +
  geom_smooth(aes(x=Yobs, y=Ypred),method=lm, se=FALSE, col="grey") +
  theme_graphs
```

## Loadings 

```{r consensus OPLS-reg loadings, message=FALSE, warning=FALSE, fig.align='center'}
loadings <- rbind.data.frame(COPLS_res@loadings$rna, COPLS_res@loadings$lipids, COPLS_res@loadings$sphingolipids)
loadings$dataset <- c(rep("genes", nrow(COPLS_res@loadings$rna)),
                      rep("lipids", nrow(COPLS_res@loadings$lipids)),
                      rep("sphingolipids",
                          nrow(COPLS_res@loadings$sphingolipids)))
loadings$variable <- rownames(loadings)

ggplot(loadings, aes(x=p_1, y=o_1, col=dataset, label = variable)) +
  geom_point(size=2, alpha = 0.5) +
  labs(x="Predictive",
       y="Orthogonal",
       title = "Loadings plots on predictive vs orthogonal latent variables") +
  #geom_text_repel(size=2, max.overlaps = 200, segment.size=.1) +
  theme_graphs+
  ggplot2::scale_color_manual(values= omics.colors)

```

## loadings and VIP of the optimal model

```{r consensus OPLS-reg VIP loadings, message=FALSE, warning=FALSE, fig.align='center'}
VIP <- data.frame(VIP = c(COPLS_res@VIP$rna, COPLS_res@VIP$lipids, COPLS_res@VIP$sphingolipids), 
                  variable = c(rownames(COPLS_res@VIP$rna), rownames(COPLS_res@VIP$lipids), rownames(COPLS_res@VIP$sphingolipids)))

loadings_VIP <- merge(loadings, VIP, by="variable")
loadings_VIP$label <- ifelse(loadings_VIP$VIP > 1, loadings_VIP$variable, NA)

ggplot(loadings_VIP, aes(x=p_1, y=VIP, col=dataset, label = label)) +
  geom_point(size=2, alpha = 0.5) +
  labs(x="Predictive",
       y="VIP",
       title = "loadings plots on predictive vs orthogonal latent variables") +
  geom_text_repel(size=3, max.overlaps = 50, segment.size=.1) +
  theme_graphs+
  ggplot2::scale_color_manual(values= omics.colors)

loadings_VIP <- loadings_VIP[order(loadings_VIP$VIP, decreasing = T),]
loadings_VIP$rank <- seq(1, nrow(loadings_VIP))
loadings_VIP <- loadings_VIP[loadings_VIP$VIP > quantile(loadings_VIP$VIP, 0.75, names = F),]

ggplot(loadings_VIP, aes(x=rank, y=VIP, fill=dataset, label = label)) +
  geom_bar(stat = "identity") +
  labs(x="Variable",
       y="VIP",
       title = "VIP plot") +
  theme_graphs+
  ggplot2::scale_color_manual(values= omics.colors)

```

```{r extract_VIP, class.source='fold-hide'}
# Compute the VIP
position <- COPLS_res@cv$nOcompOpt
VIP <- COPLS_res@VIP

# Multiply VIP * sign(loadings for predictive component)
VIP_plot <- lapply(X = 1:nbrBlocs,
                   FUN = function(X){
                       signe_loadings <- sign(COPLS_res@loadings[[X]][, "p_1"])
                       result <- VIP[[X]][, "p_1"]*signe_loadings
                       return(sort(result, decreasing = TRUE))})
names(VIP_plot) <- names(VIP)
```

```{r plot_VIP, class.source='fold-hide'}
# Rna data
data <- data.frame(
  "variables" = factor(names(VIP_plot[[1]]),
                       levels=names(VIP_plot[[1]])[order(abs(VIP_plot[[1]]), 
                                                         decreasing=T)]), 
  "valeur" = VIP_plot[[1]])
data_reduced <- data[c(1:10, (nrow(data)-9):nrow(data)), ]
VIP_rna <- ggplot2::ggplot(data = data_reduced, 
                           aes(x = variables, y = valeur)) +
  ggplot2::geom_bar(stat = "identity", fill = "#BBBBBB") +
  ggplot2::labs(title = paste0("Top ", nrow(data_reduced), " most contributed ",
                               names(VIP_plot)[1], " variables")) +
  ggplot2::xlab("Predictive variables") +
  ggplot2::ylab("VIP x loading sign") +
  theme_graphs +
  ggplot2::theme(axis.text.x = element_text(angle = 66, vjust = 1, hjust = 1)) 
VIP_rna
#plotly::ggplotly(VIP_rna)
rm(VIP_rna)

# Lipids data
data = data.frame(
  "variables" = factor(names(VIP_plot)[[2]],
                       levels=names(VIP_plot)[[2]][order(abs(VIP_plot[[2]]), 
                                                         decreasing=T)]), 
  "valeur" = VIP_plot[[2]])
data_reduced <- data[c(1:10, (nrow(data)-9):nrow(data)), ]
VIP_lipids <-  ggplot2::ggplot(data = data_reduced, 
                               aes(x = variables, y = valeur)) +
  ggplot2::geom_bar(stat = "identity", fill = "#BBBBBB") +
  ggplot2::labs(title = paste0("Top ", nrow(data_reduced), " most contributed ",
                               names(VIP_plot)[2], " variables")) +
  ggplot2::xlab("Predictive variables") +
  ggplot2::ylab("VIP x loading sign") +
  theme_graphs +
  ggplot2::theme(axis.text.x = element_text(angle = 66, vjust = 1, hjust = 1)) 
VIP_lipids
#plotly::ggplotly(VIP_lipids)
rm(VIP_lipids)

# Sphingolipids data
data = data.frame(
  "variables" = factor(names(VIP_plot)[[3]],
                       levels=names(VIP_plot)[[3]][order(abs(VIP_plot[[3]]), 
                                                         decreasing=T)]), 
  "valeur" = VIP_plot[[3]])
VIP_sphingolipids <-  ggplot2::ggplot(data = data, 
                                      aes(x = variables, y = valeur)) +
  ggplot2::geom_bar(stat = "identity", fill = "#BBBBBB") +
  ggplot2::labs(title = paste0("Top ", nrow(data), " most contributed ",
                               names(VIP_plot)[3], " variables")) +
  ggplot2::xlab("Predictive variables") +
  ggplot2::ylab("VIP x loading sign") +
  theme_graphs +
  ggplot2::theme(axis.text.x = element_text(angle = 66, vjust = 1, hjust = 1)) 
VIP_sphingolipids
#plotly::ggplotly(VIP_sphingolipids)
rm(VIP_sphingolipids)
```


# Consensus OPLS discriminant analysis of the 3 diabetes status (IGT, T2D and T3D)

```{r consensusOPLS_DA, message=FALSE, warning=FALSE, fig.align='center'}
diab.patients <- metadata$`Patient ID`[metadata$`Diabetes status` %in% c("IGT", "T2D", "T3D")]
metadata.diab <- metadata[diab.patients,]

COPLS_data_DA <- list(rna=as.matrix(rna[diab.patients,]),
                      lipids=as.matrix(lip[diab.patients,]),
                      sphingolipids = as.matrix(sphingo[diab.patients,]))
COPLS_data_DA$lipids <- apply(COPLS_data_DA$lipids, 2, as.numeric)
COPLS_data_DA <- lapply(COPLS_data_DA, scale)

Y_DA <- as.matrix(data.frame(IGT = ifelse(metadata.diab$`Diabetes status` == "IGT", 1, 0),
                             T2D = ifelse(metadata.diab$`Diabetes status` == "T2D", 1, 0),
                             T3D = ifelse(metadata.diab$`Diabetes status` == "T3D", 1, 0)))
rownames(Y_DA) <- diab.patients

COPLS_res_DA <- ConsensusOPLS(
  data = COPLS_data_DA,
  Y = Y_DA,
  maxPcomp = 2,
  maxOcomp = 3,
    modelType = "da",
    nperm = 200,
  cvType = "nfold",
  nfold = 15,
  kernelParams = list(type = "p", 
                    params = c(order = 1)),
  mc.cores = 1
)
```


## Permutations

### Main results

```{r DA_permutations, message=FALSE, warning=FALSE, fig.align='center'}
Q2perm_DA <- data.frame(Q2perm = COPLS_res_DA@permStats$Q2Y)

ggplot(data = Q2perm_DA, aes(x = Q2perm)) +
  geom_histogram(aes(y = after_stat(density)), bins = 30,
                 color="grey", fill="grey") +
  geom_density(color = "blue", linewidth = 0.5) +
  geom_vline(aes(xintercept=COPLS_res_DA@permStats$Q2Y[2]),
             color="blue", linetype="dashed", size=1) +
  theme_graphs +
  ggtitle("Q2 Permutation test")

dQ2perm_DA <- data.frame(dQ2perm = COPLS_res_DA@permStats$DQ2)

ggplot(data = dQ2perm_DA, aes(x = dQ2perm)) +
  geom_histogram(aes(y = after_stat(density)), bins = 30,
                 color="grey", fill="grey") +
  geom_density(color = "blue", linewidth = 0.5) +
  geom_vline(aes(xintercept=COPLS_res_DA@permStats$DQ2[2]),
             color="blue", linetype="dashed", size=1) +
  theme_graphs +
  ggtitle("dQ2 Permutation test")


R2Yperm_DA <- data.frame(R2Yperm_DA = COPLS_res_DA@permStats$R2Y)

ggplot(data = R2Yperm_DA, aes(x = R2Yperm_DA)) +
  geom_histogram(aes(y = after_stat(density)), bins = 30,
                 color="grey", fill="grey") +
  geom_density(color = "blue", linewidth = 0.5) +
  geom_vline(aes(xintercept=COPLS_res_DA@permStats$R2Y[2]),
             color="blue", linetype="dashed", size=1) +
  theme_graphs +
  ggtitle("R2Y Permutation test")

```

### Zoom on bimodality

```{r exclude_T3D}
diab.patients_bis <- metadata$`Patient ID`[metadata$`Diabetes status` %in% c("IGT", "T2D")]
metadata.diab_bis <- metadata[diab.patients_bis,]

COPLS_data_DA_bis <- list(rna=as.matrix(rna[diab.patients_bis,]),
                          lipids=as.matrix(lip[diab.patients_bis,]),
                          sphingolipids = as.matrix(sphingo[diab.patients_bis,]))
COPLS_data_DA_bis$lipids <- apply(COPLS_data_DA_bis$lipids, 2, as.numeric)
COPLS_data_DA_bis <- lapply(COPLS_data_DA_bis, scale)

Y_DA_bis <- as.matrix(data.frame(IGT = ifelse(metadata.diab_bis$`Diabetes status` == "IGT", 1, 0),
                                 T2D = ifelse(metadata.diab_bis$`Diabetes status` == "T2D", 1, 0)))
rownames(Y_DA_bis) <- diab.patients_bis

COPLS_res_DA_bis <- ConsensusOPLS(
  data = COPLS_data_DA_bis,
  Y = Y_DA_bis,
  maxPcomp = 1,
  maxOcomp = 3,
    modelType = "da",
    nperm = 200,
  cvType = "nfold",
  #nfold = 35,
  kernelParams = list(type = "p", 
                    params = c(order = 1)),
  mc.cores = 1
)

R2Yperm_DA <- data.frame(R2Yperm_DA = COPLS_res_DA_bis@permStats$R2Y)

ggplot(data = R2Yperm_DA, aes(x = R2Yperm_DA)) +
  geom_histogram(aes(y = after_stat(density)), bins = 30,
                 color="grey", fill="grey") +
  geom_density(color = "blue", linewidth = 0.5) +
  geom_vline(aes(xintercept=COPLS_res_DA_bis@permStats$R2Y[2]),
             color="blue", linetype="dashed", size=1) +
  theme_graphs +
  ggtitle("R2Y Permutation test")

scores_DA_bis <- data.frame(metadata.diab_bis, COPLS_res_DA_bis@scores)

ggplot(scores_DA_bis, aes(x=p_1, y=o_1, col = Diabetes.status)) +
  geom_point(size=4) +
  labs(x="Predictive 1",
       y="Orthogonal 1",
       title = "scores plots on predictive latent variables") +
  scale_color_manual(values= diabetes.status.colors) +
  theme_graphs
```

```{r exclude_T2D}
diab.patients_bis <- metadata$`Patient ID`[metadata$`Diabetes status` %in% c("IGT", "T3D")]
metadata.diab_bis <- metadata[diab.patients_bis,]

COPLS_data_DA_bis <- list(rna=as.matrix(rna[diab.patients_bis,]),
                          lipids=as.matrix(lip[diab.patients_bis,]),
                          sphingolipids = as.matrix(sphingo[diab.patients_bis,]))
COPLS_data_DA_bis$lipids <- apply(COPLS_data_DA_bis$lipids, 2, as.numeric)
COPLS_data_DA_bis <- lapply(COPLS_data_DA_bis, scale)

Y_DA_bis <- as.matrix(data.frame(IGT = ifelse(metadata.diab_bis$`Diabetes status` == "IGT", 1, 0),
                                 T3D = ifelse(metadata.diab_bis$`Diabetes status` == "T3D", 1, 0)))
rownames(Y_DA_bis) <- diab.patients_bis

COPLS_res_DA_bis <- ConsensusOPLS(
  data = COPLS_data_DA_bis,
  Y = Y_DA_bis,
  maxPcomp = 1,
  maxOcomp = 3,
    modelType = "da",
    nperm = 200,
  cvType = "nfold",
  #nfold = 33,
  kernelParams = list(type = "p", 
                    params = c(order = 1)),
  mc.cores = 1
)

R2Yperm_DA <- data.frame(R2Yperm_DA = COPLS_res_DA_bis@permStats$R2Y)

ggplot(data = R2Yperm_DA, aes(x = R2Yperm_DA)) +
  geom_histogram(aes(y = after_stat(density)), bins = 30,
                 color="grey", fill="grey") +
  geom_density(color = "blue", linewidth = 0.5) +
  geom_vline(aes(xintercept=COPLS_res_DA_bis@permStats$R2Y[2]),
             color="blue", linetype="dashed", size=1) +
  theme_graphs +
  ggtitle("R2Y Permutation test")

scores_DA_bis <- data.frame(metadata.diab_bis, COPLS_res_DA_bis@scores)

ggplot(scores_DA_bis, aes(x=p_1, y=o_1, col = Diabetes.status)) +
  geom_point(size=4) +
  labs(x="Predictive 1",
       y="Orthogonal 1",
       title = "scores plots on predictive latent variables") +
  scale_color_manual(values= diabetes.status.colors) +
  theme_graphs
```

```{r exclude_IGT}
diab.patients_bis <- metadata$`Patient ID`[metadata$`Diabetes status` %in% c("T2D", "T3D")]
metadata.diab_bis <- metadata[diab.patients_bis,]

COPLS_data_DA_bis <- list(rna=as.matrix(rna[diab.patients_bis,]),
                          lipids=as.matrix(lip[diab.patients_bis,]),
                          sphingolipids = as.matrix(sphingo[diab.patients_bis,]))
COPLS_data_DA_bis$lipids <- apply(COPLS_data_DA_bis$lipids, 2, as.numeric)
COPLS_data_DA_bis <- lapply(COPLS_data_DA_bis, scale)

Y_DA_bis <- as.matrix(data.frame(T2D = ifelse(metadata.diab_bis$`Diabetes status` == "T2D", 1, 0),
                                 T3D = ifelse(metadata.diab_bis$`Diabetes status` == "T3D", 1, 0)))
rownames(Y_DA_bis) <- diab.patients_bis

COPLS_res_DA_bis <- ConsensusOPLS(
  data = COPLS_data_DA_bis,
  Y = Y_DA_bis,
  maxPcomp = 1,
  maxOcomp = 3,
    modelType = "da",
    nperm = 200,
  cvType = "nfold",
  #nfold = 15,
  kernelParams = list(type = "p", 
                    params = c(order = 1)),
  mc.cores = 1
)

R2Yperm_DA <- data.frame(R2Yperm_DA = COPLS_res_DA_bis@permStats$R2Y)

ggplot(data = R2Yperm_DA, aes(x = R2Yperm_DA)) +
  geom_histogram(aes(y = after_stat(density)), bins = 30,
                 color="grey", fill="grey") +
  geom_density(color = "blue", linewidth = 0.5) +
  geom_vline(aes(xintercept=COPLS_res_DA_bis@permStats$R2Y[2]),
             color="blue", linetype="dashed", size=1) +
  theme_graphs +
  ggtitle("R2Y Permutation test")

scores_DA_bis <- data.frame(metadata.diab_bis, COPLS_res_DA_bis@scores)

ggplot(scores_DA_bis, aes(x=p_1, y=o_1, col = Diabetes.status)) +
  geom_point(size=4) +
  labs(x="Predictive 1",
       y="Orthogonal 1",
       title = "scores plots on predictive latent variables") +
  scale_color_manual(values= diabetes.status.colors) +
  theme_graphs
```

```{r only_IGT}
diab.patients_bis <- metadata$`Patient ID`[metadata$`Diabetes status` %in% c("IGT", "T2D", "T3D")]
metadata.diab_bis <- metadata[diab.patients_bis,]

COPLS_data_DA_bis <- list(rna=as.matrix(rna[diab.patients_bis,]),
                          lipids=as.matrix(lip[diab.patients_bis,]),
                          sphingolipids = as.matrix(sphingo[diab.patients_bis,]))
COPLS_data_DA_bis$lipids <- apply(COPLS_data_DA_bis$lipids, 2, as.numeric)
COPLS_data_DA_bis <- lapply(COPLS_data_DA_bis, scale)

Y_DA_bis <- as.matrix(data.frame(IGT = ifelse(metadata.diab_bis$`Diabetes status` == "IGT", 1, 0)))
rownames(Y_DA_bis) <- diab.patients_bis

COPLS_res_DA_bis <- ConsensusOPLS(
  data = COPLS_data_DA_bis,
  Y = Y_DA_bis,
  maxPcomp = 1,
  maxOcomp = 3,
    modelType = "da",
    nperm = 200,
  cvType = "nfold",
  #nfold = 15,
  kernelParams = list(type = "p", 
                    params = c(order = 1)),
  mc.cores = 1
)

R2Yperm_DA <- data.frame(R2Yperm_DA = COPLS_res_DA_bis@permStats$R2Y)

ggplot(data = R2Yperm_DA, aes(x = R2Yperm_DA)) +
  geom_histogram(aes(y = after_stat(density)), bins = 30,
                 color="grey", fill="grey") +
  geom_density(color = "blue", linewidth = 0.5) +
  geom_vline(aes(xintercept=COPLS_res_DA_bis@permStats$R2Y[2]),
             color="blue", linetype="dashed", size=1) +
  theme_graphs +
  ggtitle("R2Y Permutation test")

scores_DA_bis <- data.frame(metadata.diab_bis, COPLS_res_DA_bis@scores)

ggplot(scores_DA_bis, aes(x=p_1, y=o_1, col = Diabetes.status)) +
  geom_point(size=4) +
  labs(x="Predictive 1",
       y="Orthogonal 1",
       title = "scores plots on predictive latent variables") +
  scale_color_manual(values= diabetes.status.colors) +
  theme_graphs
```

```{r only_T2D}
diab.patients_bis <- metadata$`Patient ID`[metadata$`Diabetes status` %in% c("IGT", "T2D", "T3D")]
metadata.diab_bis <- metadata[diab.patients_bis,]

COPLS_data_DA_bis <- list(rna=as.matrix(rna[diab.patients_bis,]),
                          lipids=as.matrix(lip[diab.patients_bis,]),
                          sphingolipids = as.matrix(sphingo[diab.patients_bis,]))
COPLS_data_DA_bis$lipids <- apply(COPLS_data_DA_bis$lipids, 2, as.numeric)
COPLS_data_DA_bis <- lapply(COPLS_data_DA_bis, scale)

Y_DA_bis <- as.matrix(data.frame(T2D = ifelse(metadata.diab_bis$`Diabetes status` == "T2D", 1, 0)))
rownames(Y_DA_bis) <- diab.patients_bis

COPLS_res_DA_bis <- ConsensusOPLS(
  data = COPLS_data_DA_bis,
  Y = Y_DA_bis,
  maxPcomp = 1,
  maxOcomp = 3,
    modelType = "da",
    nperm = 200,
  cvType = "nfold",
  #nfold = 15,
  kernelParams = list(type = "p", 
                    params = c(order = 1)),
  mc.cores = 1
)

R2Yperm_DA <- data.frame(R2Yperm_DA = COPLS_res_DA_bis@permStats$R2Y)

ggplot(data = R2Yperm_DA, aes(x = R2Yperm_DA)) +
  geom_histogram(aes(y = after_stat(density)), bins = 30,
                 color="grey", fill="grey") +
  geom_density(color = "blue", linewidth = 0.5) +
  geom_vline(aes(xintercept=COPLS_res_DA_bis@permStats$R2Y[2]),
             color="blue", linetype="dashed", size=1) +
  theme_graphs +
  ggtitle("R2Y Permutation test")

scores_DA_bis <- data.frame(metadata.diab_bis, COPLS_res_DA_bis@scores)

ggplot(scores_DA_bis, aes(x=p_1, y=o_1, col = Diabetes.status)) +
  geom_point(size=4) +
  labs(x="Predictive 1",
       y="Orthogonal 1",
       title = "scores plots on predictive latent variables") +
  scale_color_manual(values= diabetes.status.colors) +
  theme_graphs
```

```{r only_T3D}
diab.patients_bis <- metadata$`Patient ID`[metadata$`Diabetes status` %in% c("IGT", "T2D", "T3D")]
metadata.diab_bis <- metadata[diab.patients_bis,]

COPLS_data_DA_bis <- list(rna=as.matrix(rna[diab.patients_bis,]),
                          lipids=as.matrix(lip[diab.patients_bis,]),
                          sphingolipids = as.matrix(sphingo[diab.patients_bis,]))
COPLS_data_DA_bis$lipids <- apply(COPLS_data_DA_bis$lipids, 2, as.numeric)
COPLS_data_DA_bis <- lapply(COPLS_data_DA_bis, scale)

Y_DA_bis <- as.matrix(data.frame(T3D = ifelse(metadata.diab_bis$`Diabetes status` == "T3D", 1, 0)))
rownames(Y_DA_bis) <- diab.patients_bis

COPLS_res_DA_bis <- ConsensusOPLS(
  data = COPLS_data_DA_bis,
  Y = Y_DA_bis,
  maxPcomp = 1,
  maxOcomp = 3,
    modelType = "da",
    nperm = 200,
  cvType = "nfold",
  #nfold = 15,
  kernelParams = list(type = "p", 
                    params = c(order = 1)),
  mc.cores = 1
)

R2Yperm_DA <- data.frame(R2Yperm_DA = COPLS_res_DA_bis@permStats$R2Y)

ggplot(data = R2Yperm_DA, aes(x = R2Yperm_DA)) +
  geom_histogram(aes(y = after_stat(density)), bins = 30,
                 color="grey", fill="grey") +
  geom_density(color = "blue", linewidth = 0.5) +
  geom_vline(aes(xintercept=COPLS_res_DA_bis@permStats$R2Y[2]),
             color="blue", linetype="dashed", size=1) +
  theme_graphs +
  ggtitle("R2Y Permutation test")

scores_DA_bis <- data.frame(metadata.diab_bis, COPLS_res_DA_bis@scores)

ggplot(scores_DA_bis, aes(x=p_1, y=o_1, col = Diabetes.status)) +
  geom_point(size=4) +
  labs(x="Predictive 1",
       y="Orthogonal 1",
       title = "scores plots on predictive latent variables") +
  scale_color_manual(values= diabetes.status.colors) +
  theme_graphs
```





## Blocks contribution


```{r consensus OPLS-DA contributions, message=FALSE, warning=FALSE, fig.align='center'}
contributions_DA <- COPLS_res_DA@blockContribution
contributions_DA <- melt(contributions_DA)
colnames(contributions_DA) <- c("dataset", "Dim", "value")

ggplot(contributions_DA, aes(x=Dim, y=value, fill=dataset)) +
  geom_bar(stat = "identity", position=position_dodge()) +
  theme_graphs +
  labs(x = "Global components", y = "Specific weights of block \n on global components", fill = "omic",
       title = "Blocks contributions") +
  scale_x_discrete(labels = c("Pred #1", "Pred #2", 
                              "Ortho #1", "Ortho #2"))+
  ggplot2::scale_color_manual(values= diabetes.status.colors)

```

## Scores Plot

```{r consensus OPLS-DA scores, message=FALSE, warning=FALSE, fig.align='center'}
scores_DA <- data.frame(metadata.diab, COPLS_res_DA@scores)

ggplot(scores_DA, aes(x=p_1, y=p_2, col = Diabetes.status)) +
  geom_point(size=4) +
  labs(x="Predictive 1",
       y="Predictive 2",
       title = "scores plots on predictive latent variables") +
  scale_color_manual(values= diabetes.status.colors) +
  theme_graphs
```

## Loadings 

```{r consensus OPLS-DA loadings, message=FALSE, warning=FALSE, fig.align='center'}
loadings_DA <- rbind.data.frame(COPLS_res_DA@loadings$rna, COPLS_res_DA@loadings$lipids, COPLS_res_DA@loadings$sphingolipids)
loadings_DA$dataset <- c(rep("genes", nrow(COPLS_res_DA@loadings$rna)),
                      rep("lipids", nrow(COPLS_res_DA@loadings$lipids)),
                      rep("sphingolipids", nrow(COPLS_res_DA@loadings$sphingolipids)))
loadings_DA$variable <- rownames(loadings_DA)

ggplot(loadings_DA, aes(x=p_1, y=o_1, col=dataset, label = variable)) +
  geom_point(size=2, alpha = 0.5) +
  labs(x="Predictive 1",
       y="Predictive 2",
       title = "loadings plots on predictive latent variables") +
  geom_text_repel(size=2, max.overlaps = 100, segment.size=.1) +
  theme_graphs+
  ggplot2::scale_color_manual(values= diabetes.status.colors)

```

## VIP of the optimal model

```{r consensus OPLS-DA VIP loadings, message=FALSE, warning=FALSE, fig.align='center'}
loadings <- do.call(rbind.data.frame, COPLS_res_DA@loadings)
loadings$block <- do.call(c, lapply(names(COPLS_res_DA@loadings), function(x) 
    rep(x, nrow(COPLS_res_DA@loadings[[x]]))))
loadings$variable <- gsub(paste(paste0(names(COPLS_res_DA@loadings), '.'), 
                                collapse='|'), '', 
                          rownames(loadings))

VIP <- do.call(rbind.data.frame, COPLS_res_DA@VIP)
VIP$block <- do.call(c, lapply(names(COPLS_res_DA@VIP), function(x) 
    rep(x, nrow(COPLS_res_DA@VIP[[x]]))))
VIP$variable <- gsub(paste(paste0(names(COPLS_res_DA@VIP), '.'), 
                                collapse='|'), '', 
                          rownames(VIP))
        
loadings_VIP <- merge(x = loadings[, c("p_1", "variable")], 
                      y = VIP[, c("p_1", "variable")], 
                      by = "variable", all = TRUE)
colnames(loadings_VIP) <- c("variable", "loadings", "VIP")
loadings_VIP <- merge(x = loadings_VIP, 
                      y = loadings[, c("block", "variable")], 
                      by = "variable", all = TRUE)
loadings_VIP$label <- ifelse(loadings_VIP$VIP > 1, loadings_VIP$variable, NA)



ggplot(data = loadings_VIP,
       aes(x=loadings, y=VIP, col=block, label = label)) +
  geom_point(size = 2.5, alpha = 0.5) + 
  xlab("Predictive component") +
  ylab("Variable Importance in Projection") +
  ggtitle("VIP versus loadings on predictive components")+
  theme_graphs+
  scale_color_manual(values = diabetes.status.colors)
```




